keras_cv/models/segmentation/deeplab.py

```diff
@@ -82,110 +97,95 @@ def __init__(
-                    ],
-                    include_rescaling=include_rescaling,
-                    include_top=False,
-                    name="resnet50v2",
-                    weights=parse_weights(backbone_weights, False, "resnet50v2"),
-                    pooling=None,
-                    **kwargs,
+                if backbone_weights and input_shape[-1] == 1:
+                    raise ValueError(
+                        "The input shape is set up for greyscale images with one channel, but backbone weights are trained on colored images and cannot be loaded."
+                    )
+                backbone = get_resnet_backbone(
+                    backbone_weights, include_rescaling, input_shape, **kwargs
                 )
 
         else:
             # TODO(scottzhu): Might need to do more assertion about the model
+            # What else do we want to test for? Shapes? This feels like too little, but
+            # more assertions feel like they'd be limiting.
             if not isinstance(backbone, tf.keras.layers.Layer):
                 raise ValueError(
                     "Backbone need to be a `tf.keras.layers.Layer`, "
                     f"received {backbone}"
                 )
-        self.backbone = backbone
 
+        feature_map = backbone(x)
         if spatial_pyramid_pooling is None:
-            self.aspp = SpatialPyramidPooling(dilation_rates=[6, 12, 18])
-        else:
-            self.aspp = spatial_pyramid_pooling
+            spatial_pyramid_pooling = SpatialPyramidPooling(dilation_rates=[6, 12, 18])
+
+        output = spatial_pyramid_pooling(feature_map)
+        output = tf.keras.layers.UpSampling2D(
+            size=(height // feature_map.shape[1], width // feature_map.shape[2]),
+            interpolation="bilinear",
+        )(output)
 
-        self._segmentation_head_passed = segmentation_head
         if segmentation_head is None:
-            segmentation_head = tf.keras.Sequential(
-                [
-                    tf.keras.layers.Conv2D(
-                        filters=256,
-                        kernel_size=(1, 1),
-                        padding="same",
-                        use_bias=False,
-                    ),
-                    tf.keras.layers.BatchNormalization(),
-                    tf.keras.layers.Activation("relu"),
-                    tf.keras.layers.Dropout(0.2),
-                    tf.keras.layers.Conv2D(
-                        filters=classes,
-                        kernel_size=(1, 1),
-                        padding="same",
-                        use_bias=False,
-                        activation="softmax",
-                        # Force the dtype of the classification head to float32 to avoid the NAN loss
-                        # issue when used with mixed precision API.
-                        dtype=tf.float32,
-                    ),
-                ]
+            segmentation_head = SegmentationHead(
```

Comment:
I think I intentionally moved away from __internal__.segmentation_head.py, because the implementation here works better (in terms of model quality).
Would you mind simply extracting this out, but still keep the SegmentationHead inside the file? Just similar to class RCNNHead inside faster_rcnn.py

Reply:
Yes - the segmentation head in DeepLab was different than the `SegmentationHead`, and seemed to be tailored to the model.

The reason I updated the `SegmentationHead` layer with new arguments was to make sure that it can be used to simulate the exact internal head you built in the DeepLab class. Then, I used the updated head and customized it to create an exact replica of the tailored head, sans the `use_bias` arguments (which lead to a few unnecessary params).

Generalizing the segmentation head and making it customizable is IMO a really important part of shipping sem seg in KCV.

Reply:
I'm fine with making it more generic. However do note that we're not offering segmentation head as a public API, so:
1) please extract this out from the model init section, but still keep the SegmentationHead in the same file
2) remove __internal__/segmentation_head.py file.

The reason we're doing it this way is to give users a comprehensive view of how they should build each component to work with the meta architecture. This especially applies to head where it's only a mixture of conv and dense layers.

Reply:
Hmm, I see. I misunderstood that the seghead file will eventually become a public API.
Makes sense, refactoring. Thanks for the clarification!

Reply:
Do I remove the `segmentation_head_test.py` as well?

Reply:
Did a `git mv keras_cv/models/segmentation/__internal__/segmentation_head_test.py keras_cv/models/segmentation/segmentation_head_test.py` and replaced the import so the test still works. Though, GH now treats the tests as my LOC instead of Scott's...

Looking into how I can preserve the right credit

Reply:
No need to test for the head.